src/main/java/com/iota/iri/model/Hash.java

```diff
@@ -19,33 +21,56 @@
     private byte[] bytes;
     private int[] trits;
     private int hashCode;
+	
```

Comment:
have you tried making these 3 properties final, calculated when the object is created.
The best way to make code thread safe is to make it immutable. If you need to change it in any way, just create a new instance, the functional approach is way more robust and easier to parallelize.

In my experience, `synchronized` should be a last resource and usually a sign of something bad in design. 
Try to work it functionally, with no mutability.
All those synchronized blocks most likely will end up creating deadlock headaches.

Reply:
hi tx for your suggestion.....

like u said maybe setting all properties in the constructor and make them final can make this whole class thread-safe.....of course there is no lazy initialization but at least coding-wise a lot simpler....

but there is already another PR as a possible fix https://github.com/iotaledger/iri/pull/577
..... it features lazy initialization but it still needs some sync blocks though....so all in all a bit more complex

Reply:
wish I had more time to try and code this myself immutably, perhaps a few cascading refactors might show weaknesses in the architecture